Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,290 @@
+```
+dotnet --roll-forward LatestMinor app.dll
+```
+
+For the framework references specified by the application itself:  
+The order above defines precedence. Later scopes take precedence over earlier ones. So command line wins over everything, environment variable wins over `.runtimeconfig.json` and per-framework setting wins over config-wide setting.
+
+For dependent frameworks:
+Each dependent framework (for example FX1) is loaded with an effective roll-forward value (the one used to resolve that framework). If FX1 has additional framework references those will be loaded using this same roll-forward value as the default, that is unless FX1 specifies different value in its `.runtimeconfig.json` or there's a higher precedence override (environment variable `DOTNET_ROLL_FORWARD` or command line argument `--roll-forward`).
+
+
+## Pre-release versions
+Pre-release version is a version which has a pre-release part, for example `3.0.0-preview4-27415-15`. Everything after the `-` (dash) character is a pre-release identifier. Per [semantic versioning rules](https://semver.org/) pre-release versions are ordered before the same version without any pre-release part. So `3.0.0-preview4-27415-15` comes before `3.0.0`.
+
+### Old behavior
+Before .NET Core 3.0 (so 2.2 and older) the roll-forward behavior for pre-release versions ignored any of the roll-forward related settings, that is both `rollForwardOnNoCandidateFx` as well as `applyPatches`.
+The behavior was that:
+* Release version will never roll forward to pre-release version. So for example `2.1.0` will not roll forward to `3.0.0-preview4-27415-15` even if `Major` roll-forward is allowed (same for all the other settings of `rollForwardOnNoCandidateFx` as well as `applyPatches`).
+* Pre-release version will never roll forward to release version. So for example `3.0.0-preview4-27415-15` will not roll forward to `3.0.0`. Also pre-release will only roll forward to the same `major.minor.patch`. So for example `3.0.0-preview4-27415-15` will not roll forward to `3.0.1-preview1-29000-0`.
+
+### New behavior
+The newly proposed behavior is to treat pre-release versions no different from release versions. This would mean that pre-release can roll forward to release but also that release can roll forward to pre-release. For the most part this improves the previous behavior (except when `Disable` is active in which case no roll-forward will happen):
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.0`.
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.1-preview1-29000-0`.
+
+But it also introduces cases which are somewhat controversial:
+* On `Minor` (the default) `3.0.0` would roll forward to `3.0.1-preview1-29000-0` even if `3.0.1` is available on the machine. This behavior is typically not expected.
+* On `LatestMajor` `2.0.0` would roll forward to `3.0.1-preview1-29000-0` even when `3.0.0` is available on the machine.
+
+Pros:
+* Consistent behavior across all versions
```

Comment:
There's a pro that not just about theoretical consistency and that's testing patch previews completely before they go out as RTM.

If bits behave differently on release day, there can be gaps in coverage. Today, we have a test step hack that renames the directory. That barely works for automation, but is not something we can ask people to do manually to fully dogfood / manually test an upcoming release.

If we change this proposal, I would like to at least have an opt in env var to treat previews no differently than release.

Reply:
Added a pro with this scenario - thanks!